File: server/physical_pose.go
package main

import (
	_ "embed"
	"encoding/json"
	"math"
)

//go:embed physical-pose-data.json
var physicalPoseJSON []byte
var physicalPoseData struct {
	Groups   map[string][][]int
	Controls map[string][]int
	Offsets  [][2]int
	Origins  [][2]int
}

// distance_to_point includes the querying player's vertical sprite origin.
func playerPointDistance(p *Player, x, z float64) int {
	origin := physicalPoseData.Origins[p.physicalSprite]
	return referenceDistance(p.X-float64(origin[1])*terrainUnit-x, p.Z-z)
}
func opponentDistance(p, q *Player) int {
	target, query := p, q
	if p.Team != 0 {
		target, query = q, p
	}
	origin := physicalPoseData.Origins[target.physicalSprite]
	return playerPointDistance(query, target.X-float64(origin[1])*terrainUnit, target.Z+float64(origin[0])*terrainUnit)
}

func init() {
	if err := json.Unmarshal(physicalPoseJSON, &physicalPoseData); err != nil {
		panic(err)
	}
}
func advancePhysicalPose(p *Player, i, period int, dt float64) {
	direction := (int(math.Round(math.Atan2(p.FZ, p.FX)*4/math.Pi)) + 8) % 8
	up := (p.Team == 0) == (period != 2)
	keeper := i%9 == 0
	kind := p.Action
	if kind == 0 && (p.moveX != 0 || p.moveZ != 0) {
		kind = 5
	}
	group := []string{"standing", "slide", "jump", "throw", "fall", "run", "catch", "punch"}[kind]
	index := 0
	if kind == 5 {
		index = int(math.Floor(p.poseCursor+1e-7)) & 7
		p.poseCursor = math.Mod(p.poseCursor+dt*25, 8)
	} else if kind != 0 {
		duration := 4. / 25
		switch kind {
		case 1, 2:
			duration = actionDuration(kind, p.Stats[3])
		case 3:
			duration = 8. / 25
		case 4:
			duration = fallDuration
		case 6:
			duration = 3. / 25
		}
		if p.poseKind == kind && p.ActionTime <= p.poseRemaining+1e-9 {
			duration = p.poseDuration
		}
		p.poseDuration = duration
		index = max(0, int(math.Floor((duration-p.ActionTime)*25+1e-7)))
		if kind == 2 && p.ActionTime <= 2./25+1e-9 {
			index = 18 + max(0, int(math.Floor((2./25-p.ActionTime)*25+1e-7)))
		}
		if kind == 1 && p.ActionTime <= 1./25+1e-9 {
			index = 15
		}
		p.poseCursor = float64(index + 1)
	} else {
		p.poseCursor = 0
	}
	if keeper && (kind == 0 || kind == 5 || kind == 6) {
		if kind == 0 {
			group = "stand"
		} else if kind == 5 {
			group = "run"
		} else {
			group = "catch"
		}
		if up {
			group += "Up"
		} else {
			group += "Down"
		}
	}
	if keeper && kind == 0 {
		direction = 4
		if up {
			direction = 0
		}
	}
	frames := physicalPoseData.Groups[group][direction]
	index = min(index, len(frames)-1)
	p.poseCursor = float64(index + 1)
	if index == len(frames)-1 {
		switch physicalPoseData.Controls[group][direction] {
		case -1, -4:
			p.poseCursor = 0
		case -2, -5:
			p.poseCursor = float64(index)
		}
	}
	p.physicalSprite = frames[index]
	p.physicalPoseValid = true
	p.physicalFrame = index
	p.poseKind = kind
	p.poseRemaining = p.ActionTime
}
func physicalBallOffset(p *Player, b *Ball) (float64, float64) {
	correction := 12
	if b.heldJump && p.Action == 2 && p.physicalFrame >= 2 && p.physicalFrame < 18 {
		correction = 8
	}
	origin := physicalPoseData.Offsets[p.physicalSprite]
	return -float64(origin[1]-correction) * terrainUnit, float64(origin[0]-correction) * terrainUnit
}

File: src/physical-pose.js
import data from "./physical-pose-data.json" with { type: "json" };
import {
  actionDuration,
  fallDuration,
  referenceDistance,
} from "./attributes.js";
const unit = 22.4 / 576;
export function playerTargetDelta(p, target) {
  const a = data.origins[p.physicalSprite ?? 0];
  const b = data.origins[target.physicalSprite ?? 0];
  return { x: target.x - b[1] * unit - p.x + a[1] * unit,
    z: target.z + b[0] * unit - p.z };
}
// distance_to_point (0xdaca) includes only the querying player's Y origin.
export function playerPointDistance(p, x, z) {
  const origin = data.origins[p.physicalSprite ?? 0];
  return referenceDistance(p.x - origin[1] * unit - x, p.z - z);
}
// calculate_player_distances (0xda20) adds both origins to team one's target,
// queries team two, then stores the same result for both opponents.
export function opponentDistance(p, q) {
  const target = p.team === 0 ? p : q;
  const query = p.team === 0 ? q : p;
  const origin = data.origins[target.physicalSprite ?? 0];
  return playerPointDistance(
    query,
    target.x - origin[1] * unit,
    target.z + origin[0] * unit,
  );
}
// Advance original sprite selection for physical carried-ball placement.
export function advancePhysicalPose(p, i, period, dt) {
  const direction =
    (Math.round((Math.atan2(p.fz, p.fx) * 4) / Math.PI) + 8) % 8;
  const up = (p.team === 0) === (period !== 2),
    keeper = i % 9 === 0;
  let kind = p.action || (p.moveX || p.moveZ ? 5 : 0),
    group = [
      "standing",
      "slide",
      "jump",
      "throw",
      "fall",
      "run",
      "catch",
      "punch",
    ][kind];
  let index = 0;
  if (kind === 5) {
    index = Math.floor((p.poseCursor || 0) + 1e-7) & 7;
    p.poseCursor = ((p.poseCursor || 0) + dt * 25) % 8;
  } else if (kind) {
    const computedDuration =
      kind === 1 || kind === 2
        ? actionDuration(kind, p.stats[3])
        : kind === 3
          ? 8 / 25
          : kind === 4
            ? fallDuration
            : kind === 6
              ? 3 / 25
              : 4 / 25;
    const duration =
      p.poseKind === kind && p.actionTime <= (p.poseRemaining ?? 0) + 1e-9
        ? p.poseDuration
        : computedDuration;
    p.poseDuration = duration;
    index = Math.max(0, Math.floor((duration - p.actionTime) * 25 + 1e-7));
    if (kind === 2 && p.actionTime <= 2 / 25 + 1e-9)
      index = 18 + Math.max(0, Math.floor((2 / 25 - p.actionTime) * 25 + 1e-7));
    if (kind === 1 && p.actionTime <= 1 / 25 + 1e-9) index = 15;
    p.poseCursor = index + 1;
  } else p.poseCursor = 0;
  if (keeper && [0, 5, 6].includes(kind))
    group =
      (kind === 0 ? "stand" : kind === 5 ? "run" : "catch") +
      (up ? "Up" : "Down");
  const frames =
    data.groups[group][keeper && kind === 0 ? (up ? 0 : 4) : direction];
  index = Math.min(index, frames.length - 1);
  p.physicalSprite = frames[index];
  p.physicalFrame = index;
  p.poseKind = kind;
  p.poseRemaining = p.actionTime;
}
export function physicalBallOffset(p, b) {
  const ballSprite =
    b.heldJump && p.action === 2 && p.physicalFrame >= 2 && p.physicalFrame < 18
      ? 6
      : 0;
  const origin = data.offsets[p.physicalSprite],
    correction = ballSprite === 0 ? 12 : 8;
  return {
    x: -(origin[1] - correction) * unit,
    z: (origin[0] - correction) * unit,
  };
}
